src/checkpoint/dispatch/dispatch.impl.h

```diff
@@ -59,23 +63,110 @@ inline Serializer& operator|(Serializer& s, T& target) {
 
 namespace checkpoint { namespace dispatch {
 
+// Only temporary BEGIN
+//
+
+template <typename T>
+struct DoNothing {
+  template <typename U = T>
+  static typename std::enable_if<std::is_same<U, int>::value>::type doNothing() {
+      throw std::runtime_error("err");
+   }
+
+  template <typename U = T>
+  static typename std::enable_if<not std::is_same<U, int>::value>::type doNothing() {
+    return;
+  }
+};
+
+//
+// Only temporary END
+
+template <typename T, typename TraverserT>
+TraverserT& withTypeIdx(TraverserT& t) {
+  using CleanT = typename CleanType<std::size_t>::CleanT;
+  using DispatchType =
+      typename TraverserT::template DispatcherType<TraverserT, CleanT>;
+
+  SerializerDispatch<TraverserT, CleanT, DispatchType> ap;
+  auto const thisTypeIdx = vrt::typeregistry::getTypeIdx<T>();
+  constexpr SerialSizeType len = 1;
+
+  if (t.isPacking() || t.isSizing()) {
+    auto val = cleanType(&thisTypeIdx);
+    ap(t, val, len);
+  } else if (t.isUnpacking()) {
+    std::size_t serTypeIdx = 0;
+    auto val = cleanType(&serTypeIdx);
+    ap(t, val, len);
+
+    // Only temporary BEGIN
+    //
+
+    try {
+      DoNothing<T>::doNothing();
+    } catch(...) {
+      auto err =
+          std::string("Unpacking wrong type, got=") +
+          vrt::typeregistry::getTypeNameForIdx(thisTypeIdx) +
+          ", expected=" + vrt::typeregistry::getTypeNameForIdx(serTypeIdx);
+      throw std::runtime_error(err);
+    }
+
+    //
+    // Only temporary END
+
+    if (thisTypeIdx != serTypeIdx) {
+      auto err =
+          std::string("Unpacking wrong type, got=") +
+          vrt::typeregistry::getTypeNameForIdx(thisTypeIdx) +
+          ", expected=" + vrt::typeregistry::getTypeNameForIdx(serTypeIdx);
+      throw std::runtime_error(err);
+    }
+  }
+
+  return t;
+}
+
 template <typename T, typename TraverserT>
 TraverserT& Traverse::with(T& target, TraverserT& t, SerialSizeType len) {
+  withTypeIdx<T>(t);
 
   using CleanT = typename CleanType<T>::CleanT;
   using DispatchType = typename TraverserT::template DispatcherType<TraverserT, CleanT>;
 
   auto val = cleanType(&target);
 
   SerializerDispatch<TraverserT, CleanT, DispatchType> ap;
-  ap(t, val, len);
+  try {
+    ap(t, val, len);
+  } catch (std::runtime_error &err) {
+    auto const what = std::string{err.what()} +
+                      "\npath: " + vrt::typeregistry::getTypeName<T>();
+    throw std::runtime_error(what);
+  }
```

Comment:
There's a lot of temporary code at the moment, but my general proposition (and IMHO the easiest one) is to simply throw from `withTypeIdx<T>()`, do a `try-catch` around `ap()` and for each catch add information about current type.

I added a temporary super simple test (`test_struct_ee`) to check this, and the result looks like this:

```
Unpacking wrong type, got=i, expected=i
path: N10checkpoint5tests4unit2AAE
path: N10checkpoint5tests4unit2BBE
path: N10checkpoint5tests4unit2CCE
path: N10checkpoint5tests4unit2DDE
path: N10checkpoint5tests4unit2EEE
```

What's left is to actually do this only if error checking is enabled. If it's not, `ap()` needs to be called without `try-catch` (and also whole `withTypeIdx()` should probably be erased).

Does it make any sense @lifflander @PhilMiller ?

Reply:
Yes, this makes sense to me. Although I'm not sure I completely understand the necessity of `DoNothing`. Can you explain why you are checking for an `int` there?

Reply:
That's just temporary code to easily test that try-catch solution. I'll remove it soon.